File: dream-server/extensions/services/dashboard-api/routers/features.py
"""Feature discovery endpoints."""

import logging
import os
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException

from config import FEATURES, GPU_BACKEND, SERVICES
from gpu import get_gpu_info, get_gpu_tier
from models import GPUInfo
from security import verify_api_key

logger = logging.getLogger(__name__)

router = APIRouter(tags=["features"])


def calculate_feature_status(feature: dict, services: list, gpu_info: Optional[GPUInfo]) -> dict:
    """Calculate whether a feature can be enabled and its status."""
    gpu_vram_gb = (gpu_info.memory_total_mb / 1024) if gpu_info else 0
    gpu_vram_used_gb = (gpu_info.memory_used_mb / 1024) if gpu_info else 0
    gpu_vram_free_gb = gpu_vram_gb - gpu_vram_used_gb

    # On Apple Silicon, when HOST_CHIP is missing (get_gpu_info_apple returned None),
    # fall back to HOST_RAM_GB. Unified memory = VRAM on Apple Silicon.
    if gpu_vram_gb == 0 and GPU_BACKEND == "apple":
        try:
            gpu_vram_gb = float(os.environ.get("HOST_RAM_GB", "0") or "0")
        except (ValueError, TypeError):
            pass
        gpu_vram_free_gb = gpu_vram_gb  # assumes zero current usage; Docker can't measure host memory pressure

    req = feature["requirements"]
    vram_ok = gpu_vram_gb >= req.get("vram_gb", 0)
    vram_fits = gpu_vram_free_gb >= req.get("vram_gb", 0)

    required_services = req.get("services", [])
    required_services_any = req.get("services_any", [])
    all_required = list(dict.fromkeys(required_services + required_services_any))
    services_available = []
    services_missing = []

    for svc_id in all_required:
        svc_status = next((s for s in services if s.id == svc_id), None)
        if svc_status and svc_status.status == "healthy":
            services_available.append(svc_id)
        else:
            services_missing.append(svc_id)

    services_all_ok = all(svc in services_available for svc in required_services)
    services_any_ok = (not required_services_any) or any(svc in services_available for svc in required_services_any)
    services_ok = services_all_ok and services_any_ok

    enabled_all = feature.get("enabled_services_all", required_services)
    enabled_any = feature.get("enabled_services_any", required_services_any)
    enabled_all_ok = all(
        any(s.id == svc and s.status == "healthy" for s in services) for svc in enabled_all
    )
    enabled_any_ok = (not enabled_any) or any(
        any(s.id == svc and s.status == "healthy" for s in services) for svc in enabled_any
    )
    is_enabled = enabled_all_ok and enabled_any_ok

    if is_enabled:
        status = "enabled"
    elif not vram_ok:
        status = "insufficient_vram"
    elif not services_ok:
        status = "services_needed"
    else:
        status = "available"

    return {
        "id": feature["id"],
        "name": feature["name"],
        "description": feature.get("description", ""),
        "icon": feature.get("icon", "Package"),
        "category": feature.get("category", "other"),
        "status": status,
        "enabled": is_enabled,
        "requirements": {
            "vramGb": req.get("vram_gb", 0),
            "vramOk": vram_ok,
            "vramFits": vram_fits,
            "services": all_required,
            "servicesAll": required_services,
            "servicesAny": required_services_any,
            "servicesAvailable": services_available,
            "servicesMissing": services_missing,
            "servicesOk": services_ok,
        },
        "setupTime": feature.get("setup_time", "Unknown"),
        "priority": feature.get("priority", 99)
    }


@router.get("/api/features")
async def api_features(api_key: str = Depends(verify_api_key)):
    """Get feature discovery data."""
    from helpers import get_all_services
    gpu_info = get_gpu_info()
    service_list = await get_all_services()

    feature_statuses = [calculate_feature_status(f, service_list, gpu_info) for f in FEATURES]
    feature_statuses.sort(key=lambda x: x["priority"])

    enabled_count = sum(1 for f in feature_statuses if f["enabled"])
    available_count = sum(1 for f in feature_statuses if f["status"] == "available")
    total_count = len(feature_statuses)

    suggestions = []
    for f in feature_statuses:
        if f["status"] == "available":
            suggestions.append({
                "featureId": f["id"], "name": f["name"],
                "message": f"Your hardware can run {f['name']}. Enable it?",
                "action": f"Enable {f['name']}", "setupTime": f["setupTime"]
            })
        elif f["status"] == "services_needed":
            missing = ", ".join(f["requirements"]["servicesMissing"])
            suggestions.append({
                "featureId": f["id"], "name": f["name"],
                "message": f"{f['name']} needs {missing} to be running.",
                "action": f"Start {missing}", "setupTime": f["setupTime"], "blocked": True
            })

    gpu_vram_gb = (gpu_info.memory_total_mb / 1024) if gpu_info else 0
    memory_type = gpu_info.memory_type if gpu_info else "discrete"

    # Apply Apple Silicon fallback for endpoint-level GPU summary (mirrors calculate_feature_status)
    if gpu_vram_gb == 0 and GPU_BACKEND == "apple":
        try:
            gpu_vram_gb = float(os.environ.get("HOST_RAM_GB", "0") or "0")
        except (ValueError, TypeError):
            pass
        if gpu_vram_gb == 0:
            logger.warning(
                "Apple Silicon VRAM fallback: HOST_RAM_GB is 0 or unset; "
                "all features will show insufficient_vram"
            )
        memory_type = "unified"

    tier_recommendations = []
    if memory_type == "unified" and gpu_info and gpu_info.gpu_backend == "amd":
        if gpu_vram_gb >= 90:
            tier_recommendations = ["Strix Halo 90+ — running qwen3-coder-next (80B MoE, 3B active)", "Plenty of headroom for the flagship model + bootstrap simultaneously", "Voice and Documents work alongside the LLM"]
        else:
            tier_recommendations = ["Strix Halo Compact — running qwen3:30b-a3b (30B MoE, 3B active)", "Fast MoE inference with low memory footprint", "Voice and Documents work alongside the LLM"]
    elif gpu_vram_gb >= 80:
        tier_recommendations = ["Your GPU can run all features simultaneously", "Consider enabling Voice + Documents for the full experience", "Image generation is supported at full quality"]
    elif gpu_vram_gb >= 24:
        tier_recommendations = ["Great GPU for local AI — most features will run well", "Voice and Documents work together", "Image generation may require model unloading"]
    elif gpu_vram_gb >= 16:
        tier_recommendations = ["Solid GPU for core features", "Voice works well with the default model", "For images, use a smaller chat model"]
    elif gpu_vram_gb >= 8:
        tier_recommendations = ["Entry-level GPU — focus on chat first", "Voice is possible with a smaller model", "Consider using the 7B model for better speed"]
    else:
        tier_recommendations = ["Limited GPU memory — chat will work with small models", "Consider cloud hybrid mode for better quality"]

    return {
        "features": feature_statuses,
        "summary": {"enabled": enabled_count, "available": available_count, "total": total_count, "progress": round(enabled_count / total_count * 100) if total_count > 0 else 0},
        "suggestions": suggestions[:3],
        "recommendations": tier_recommendations,
        "gpu": {"name": gpu_info.name if gpu_info else "Unknown", "vramGb": round(gpu_vram_gb, 1), "tier": get_gpu_tier(gpu_vram_gb, memory_type)}
    }


@router.get("/api/features/{feature_id}/enable")
async def feature_enable_instructions(feature_id: str, api_key: str = Depends(verify_api_key)):
    """Get instructions to enable a specific feature."""
    feature = next((f for f in FEATURES if f["id"] == feature_id), None)
    if not feature:
        raise HTTPException(status_code=404, detail=f"Feature not found: {feature_id}")

    def _svc_url(service_id: str) -> str:
        cfg = SERVICES.get(service_id, {})
        port = cfg.get("external_port", cfg.get("port", 0))
        return f"http://localhost:{port}" if port else ""

    def _svc_port(service_id: str) -> int:
        cfg = SERVICES.get(service_id, {})
        return cfg.get("external_port", cfg.get("port", 0))

    webui_url = _svc_url("open-webui")
    dashboard_url = _svc_url("dashboard")
    n8n_url = _svc_url("n8n")

    instructions = {
        "chat": {"steps": ["Chat is already enabled if llama-server is running", "Open the Dashboard and click 'Chat' to start"], "links": [{"label": "Open Chat", "url": webui_url}]},
        "voice": {"steps": [f"Ensure Whisper (STT) is running on port {_svc_port('whisper')}", f"Ensure Kokoro (TTS) is running on port {_svc_port('tts')}", "Start LiveKit for WebRTC", "Open the Voice page in the Dashboard"], "links": [{"label": "Voice Dashboard", "url": f"{dashboard_url}/voice"}]},
        "documents": {"steps": ["Ensure Qdrant vector database is running", "Enable the 'Document Q&A' workflow", "Upload documents via the workflow endpoint"], "links": [{"label": "Workflows", "url": f"{dashboard_url}/workflows"}]},
        "workflows": {"steps": [f"Ensure n8n is running on port {_svc_port('n8n')}", "Open the Workflows page to see available automations", "Click 'Enable' on any workflow to import it"], "links": [{"label": "n8n Dashboard", "url": n8n_url}, {"label": "Workflows", "url": f"{dashboard_url}/workflows"}]},
        "images": {"steps": ["Image generation requires additional setup", "Coming soon in a future update"], "links": []},
        "coding": {"steps": ["Switch to the Qwen2.5-Coder model for best results", "Use the model manager to download and load it", "Chat will now be optimized for code"], "links": [{"label": "Model Manager", "url": f"{dashboard_url}/models"}]},
    }

    return {"featureId": feature_id, "name": feature["name"], "instructions": instructions.get(feature_id, {"steps": [], "links": []})}

File: dream-server/extensions/services/dashboard-api/crates/dashboard-api/src/routes/features.rs
//! Features router — /api/features/* endpoints. Mirrors routers/features.py.

use axum::extract::{Path, State};
use axum::Json;
use serde_json::{json, Value};

use crate::state::AppState;

/// GET /api/features — list all loaded feature definitions
pub async fn list_features(State(state): State<AppState>) -> Json<Value> {
    Json(json!(*state.features))
}

/// GET /api/features/:feature_id/enable — instructions to enable a specific feature
pub async fn feature_enable(
    State(state): State<AppState>,
    Path(feature_id): Path<String>,
) -> Json<Value> {
    let feature = state
        .features
        .iter()
        .find(|f| f["id"].as_str() == Some(feature_id.as_str()));

    let feature = match feature {
        Some(f) => f,
        None => {
            return Json(json!({"error": format!("Feature not found: {feature_id}")}));
        }
    };

    fn svc_url(services: &std::collections::HashMap<String, dream_common::manifest::ServiceConfig>, id: &str) -> String {
        if let Some(cfg) = services.get(id) {
            let port = cfg.external_port;
            if port > 0 { return format!("http://localhost:{port}"); }
        }
        String::new()
    }

    fn svc_port(services: &std::collections::HashMap<String, dream_common::manifest::ServiceConfig>, id: &str) -> u16 {
        services.get(id).map(|c| c.external_port).unwrap_or(0)
    }

    let webui_url = svc_url(&state.services, "open-webui");
    let dashboard_url = svc_url(&state.services, "dashboard");
    let n8n_url_val = svc_url(&state.services, "n8n");

    let instructions = match feature_id.as_str() {
        "chat" => json!({"steps": ["Chat is already enabled if llama-server is running", "Open the Dashboard and click 'Chat' to start"], "links": [{"label": "Open Chat", "url": webui_url}]}),
        "voice" => json!({"steps": [format!("Ensure Whisper (STT) is running on port {}", svc_port(&state.services, "whisper")), format!("Ensure Kokoro (TTS) is running on port {}", svc_port(&state.services, "tts")), "Start LiveKit for WebRTC".to_string(), "Open the Voice page in the Dashboard".to_string()], "links": [{"label": "Voice Dashboard", "url": format!("{dashboard_url}/voice")}]}),
        "documents" => json!({"steps": ["Ensure Qdrant vector database is running", "Enable the 'Document Q&A' workflow", "Upload documents via the workflow endpoint"], "links": [{"label": "Workflows", "url": format!("{dashboard_url}/workflows")}]}),
        "workflows" => json!({"steps": [format!("Ensure n8n is running on port {}", svc_port(&state.services, "n8n")), "Open the Workflows page to see available automations".to_string(), "Click 'Enable' on any workflow to import it".to_string()], "links": [{"label": "n8n Dashboard", "url": n8n_url_val}, {"label": "Workflows", "url": format!("{dashboard_url}/workflows")}]}),
        "images" => json!({"steps": ["Image generation requires additional setup", "Coming soon in a future update"], "links": []}),
        "coding" => json!({"steps": ["Switch to the Qwen2.5-Coder model for best results", "Use the model manager to download and load it", "Chat will now be optimized for code"], "links": [{"label": "Model Manager", "url": format!("{dashboard_url}/models")}]}),
        "observability" => json!({"steps": [format!("Langfuse is running on port {}", svc_port(&state.services, "langfuse")), "Open Langfuse to view LLM traces and evaluations".to_string(), "LiteLLM automatically sends traces — no additional configuration needed".to_string()], "links": [{"label": "Open Langfuse", "url": svc_url(&state.services, "langfuse")}]}),
        _ => json!({"steps": [], "links": []}),
    };

    Json(json!({
        "featureId": feature_id,
        "name": feature["name"],
        "instructions": instructions,
    }))
}

/// GET /api/features/status — feature definitions with current health status.
pub async fn features_status(State(state): State<AppState>) -> Json<Value> {
    let cached = state.services_cache.read().await;
    let health_map: std::collections::HashMap<String, String> = cached
        .as_ref()
        .map(|statuses| {
            statuses
                .iter()
                .map(|s| (s.id.clone(), s.status.clone()))
                .collect()
        })
        .unwrap_or_default();

    let features: Vec<Value> = state
        .features
        .iter()
        .map(|f| {
            let mut feat = f.clone();
            let id = feat["id"].as_str().unwrap_or("");
            feat["health"] = json!(health_map.get(id).cloned().unwrap_or_else(|| "unknown".to_string()));
            feat
        })
        .collect();

    Json(json!(features))
}

#[cfg(test)]
mod tests {
    use axum::body::Body;
    use http::Request;
    use http_body_util::BodyExt;
    use serde_json::{json, Value};
    use std::collections::HashMap;
    use tower::ServiceExt;

    use crate::state::AppState;
    use dream_common::manifest::ServiceConfig;

    fn test_service_config(name: &str, port: u16) -> ServiceConfig {
        ServiceConfig {
            host: name.into(),
            port,
            external_port: port,
            health: "/health".into(),
            name: name.into(),
            ui_path: "/".into(),
            service_type: None,
            health_port: None,
        }
    }

    fn test_state_with_features(features: Vec<Value>) -> AppState {
        let mut services = HashMap::new();
        services.insert("open-webui".into(), test_service_config("open-webui", 3000));
        services.insert("dashboard".into(), test_service_config("dashboard", 3001));
        AppState::new(services, features, vec![], "test-key".into())
    }

    fn auth_header() -> (&'static str, &'static str) {
        ("Authorization", "Bearer test-key")
    }

    #[tokio::test]
    async fn test_list_features_returns_features_from_state() {
        let features = vec![
            json!({"id": "chat", "name": "Chat"}),
            json!({"id": "voice", "name": "Voice"}),
        ];
        let app = crate::build_router(test_state_with_features(features));

        let req = Request::builder()
            .uri("/api/features")
            .header(auth_header().0, auth_header().1)
            .body(Body::empty())
            .unwrap();
        let resp = app.oneshot(req).await.unwrap();
        assert_eq!(resp.status(), 200);

        let body = resp.into_body().collect().await.unwrap().to_bytes();
        let val: Value = serde_json::from_slice(&body).unwrap();
        let arr = val.as_array().unwrap();
        assert_eq!(arr.len(), 2);
        assert_eq!(arr[0]["id"], "chat");
        assert_eq!(arr[1]["id"], "voice");
    }

    #[tokio::test]
    async fn test_features_status_empty_cache_returns_unknown_health() {
        let features = vec![
            json!({"id": "chat", "name": "Chat"}),
            json!({"id": "voice", "name": "Voice"}),
        ];
        let app = crate::build_router(test_state_with_features(features));

        let req = Request::builder()
            .uri("/api/features/status")
            .header(auth_header().0, auth_header().1)
            .body(Body::empty())
            .unwrap();
        let resp = app.oneshot(req).await.unwrap();
        assert_eq!(resp.status(), 200);

        let body = resp.into_body().collect().await.unwrap().to_bytes();
        let val: Value = serde_json::from_slice(&body).unwrap();
        let arr = val.as_array().unwrap();
        assert_eq!(arr.len(), 2);
        // With no cached service statuses, every feature should have "unknown" health
        assert_eq!(arr[0]["health"], "unknown");
        assert_eq!(arr[1]["health"], "unknown");
    }

    #[tokio::test]
    async fn test_feature_enable_chat_returns_steps() {
        let features = vec![json!({"id": "chat", "name": "Chat"})];
        let app = crate::build_router(test_state_with_features(features));

        let req = Request::builder()
            .uri("/api/features/chat/enable")
            .header(auth_header().0, auth_header().1)
            .body(Body::empty())
            .unwrap();
        let resp = app.oneshot(req).await.unwrap();
        assert_eq!(resp.status(), 200);

        let body = resp.into_body().collect().await.unwrap().to_bytes();
        let val: Value = serde_json::from_slice(&body).unwrap();
        assert_eq!(val["featureId"], "chat");
        assert_eq!(val["name"], "Chat");

        let steps = val["instructions"]["steps"].as_array().unwrap();
        assert!(!steps.is_empty());
        // Verify the chat-specific instruction content
        assert!(steps[0].as_str().unwrap().contains("llama-server"));
    }

    #[tokio::test]
    async fn test_features_requires_auth() {
        let features = vec![json!({"id": "chat", "name": "Chat"})];
        let app = crate::build_router(test_state_with_features(features));
        let req = Request::builder()
            .uri("/api/features")
            .body(Body::empty())
            .unwrap();
        let resp = app.oneshot(req).await.unwrap();
        assert_eq!(resp.status(), 401);
    }

    #[tokio::test]
    async fn test_features_status_requires_auth() {
        let features = vec![json!({"id": "chat", "name": "Chat"})];
        let app = crate::build_router(test_state_with_features(features));
        let req = Request::builder()
            .uri("/api/features/status")
            .body(Body::empty())
            .unwrap();
        let resp = app.oneshot(req).await.unwrap();
        assert_eq!(resp.status(), 401);
    }

    #[tokio::test]
    async fn test_feature_enable_requires_auth() {
        let features = vec![json!({"id": "chat", "name": "Chat"})];
        let app = crate::build_router(test_state_with_features(features));
        let req = Request::builder()
            .uri("/api/features/chat/enable")
            .body(Body::empty())
            .unwrap();
        let resp = app.oneshot(req).await.unwrap();
        assert_eq!(resp.status(), 401);
    }

    #[tokio::test]
    async fn test_feature_enable_voice_returns_steps() {
        let features = vec![json!({"id": "voice", "name": "Voice"})];
        let app = crate::build_router(test_state_with_features(features));

        let req = Request::builder()
            .uri("/api/features/voice/enable")
            .header(auth_header().0, auth_header().1)
            .body(Body::empty())
            .unwrap();
        let resp = app.oneshot(req).await.unwrap();
        assert_eq!(resp.status(), 200);

        let body = resp.into_body().collect().await.unwrap().to_bytes();
        let val: Value = serde_json::from_slice(&body).unwrap();
        assert_eq!(val["featureId"], "voice");
        let steps = val["instructions"]["steps"].as_array().unwrap();
        assert!(!steps.is_empty());
        assert!(steps[0].as_str().unwrap().contains("Whisper"));
    }

    #[tokio::test]
    async fn test_feature_enable_workflows_includes_n8n() {
        let features = vec![json!({"id": "workflows", "name": "Workflows"})];
        let app = crate::build_router(test_state_with_features(features));

        let req = Request::builder()
            .uri("/api/features/workflows/enable")
            .header(auth_header().0, auth_header().1)
            .body(Body::empty())
            .unwrap();
        let resp = app.oneshot(req).await.unwrap();
        assert_eq!(resp.status(), 200);

        let body = resp.into_body().collect().await.unwrap().to_bytes();
        let val: Value = serde_json::from_slice(&body).unwrap();
        assert_eq!(val["featureId"], "workflows");
        let steps = val["instructions"]["steps"].as_array().unwrap();
        assert!(steps[0].as_str().unwrap().contains("n8n"));
    }

    #[tokio::test]
    async fn test_feature_enable_unknown_returns_error() {
        let features = vec![json!({"id": "chat", "name": "Chat"})];
        let app = crate::build_router(test_state_with_features(features));

        let req = Request::builder()
            .uri("/api/features/nonexistent/enable")
            .header(auth_header().0, auth_header().1)
            .body(Body::empty())
            .unwrap();
        let resp = app.oneshot(req).await.unwrap();
        assert_eq!(resp.status(), 200);

        let body = resp.into_body().collect().await.unwrap().to_bytes();
        let val: Value = serde_json::from_slice(&body).unwrap();
        let err = val["error"].as_str().unwrap();
        assert!(err.contains("Feature not found"));
        assert!(err.contains("nonexistent"));
    }
}
